router/app_cmds.c

```diff
@@ -117,11 +140,66 @@ static int handle_keep_alive(struct diag_client *client, const void *buf,
 	return dm_send(client, resp, sizeof(resp));
 }
 
+static int handle_diag_id(struct diag_client *client, const void *buf,
+				size_t len)
+{
+	struct diag_cmd_diag_id_query_req_t *req = NULL;
+	struct diag_cmd_diag_id_query_rsp_t *resp = NULL;
+	struct list_head *start;
+	struct diag_id_tbl_t *item = NULL;
+	size_t resp_size = DIAG_MAX_RSP_SIZE;
+	int num_entries = 0;
+	uint8_t process_name_len = 0;
+	uint8_t *dest_buf;
+	size_t write_len;
+
+	if (!buf || len < sizeof(struct diag_cmd_diag_id_query_req_t)) {
+		return -EMSGSIZE;
+	}
+	req = (struct diag_cmd_diag_id_query_req_t *) buf;
+	resp = alloca(resp_size);
+	if (!resp) {
+		return -EMSGSIZE;
+	}
+
+	dest_buf = (uint8_t *) resp;
+	resp->header.cmd_code = req->header.cmd_code;
+	resp->header.subsys_id = req->header.subsys_id;
+	resp->header.subsys_cmd_code = req->header.subsys_cmd_code;
+
+	resp->version = req->version;
+	resp->entry.process_name = NULL;
+	resp->entry.len = 0;
+	resp->entry.diag_id = 0;
+	write_len = sizeof(resp->header) + sizeof(resp->version) +
+			sizeof(resp->num_entries);
+
+	list_for_each(start, &g_diag_id_list) {
+		item = list_entry(start, struct diag_id_tbl_t, link);
+		memcpy(dest_buf + write_len, &item->diag_id,
+			sizeof(item->diag_id));
+		write_len = write_len + sizeof(item->diag_id);
+		process_name_len = strlen(item->process_name) + 1;
+		memcpy(dest_buf + write_len, &process_name_len,
+			sizeof(process_name_len));
+		write_len = write_len + sizeof(process_name_len);
+		memcpy(dest_buf + write_len, item->process_name,
+			strlen(item->process_name) + 1);
+		write_len = write_len + strlen(item->process_name) + 1;
+		num_entries++;
+	}
```

Comment:
sure, using a zero-length array is a better way. In addition, in the "list_for_each_entry" part of the logic, we must copy the message strictly byte by byte, otherwise QXDM will not be able to parse the response message.